code/modules/food_and_drinks/pizzabox.dm

```diff
@@ -330,28 +344,43 @@
 		pizza_preferences = list()
 
 /obj/item/pizzabox/infinite/examine(mob/user)
+	if(!open && ishuman(user))
+		attune_pizza(user) //pizza tag changes based on examiner
 	. = ..()
 	if(isobserver(user))
 		. += "<span class='deadsay'>This pizza box is anomalous, and will produce infinite pizza.</span>"
 
 /obj/item/pizzabox/infinite/attack_self(mob/living/user)
-	QDEL_NULL(pizza)
 	if(ishuman(user))
 		attune_pizza(user)
-	. = ..()
+	return ..()
+
+/obj/item/pizzabox/infinite/proc/attune_pizza(mob/living/carbon/human/nommer) //tonight on "proc names I never thought I'd type"
+	if(!nommer.ckey)
+		return
 
-/obj/item/pizzabox/infinite/proc/attune_pizza(mob/living/carbon/human/noms) //tonight on "proc names I never thought I'd type"
-	if(!pizza_preferences[noms.ckey])
-		pizza_preferences[noms.ckey] = pickweight(pizza_types)
-		if(noms.has_quirk(/datum/quirk/pineapple_liker))
-			pizza_preferences[noms.ckey] = /obj/item/food/pizza/pineapple
-		else if(noms.has_quirk(/datum/quirk/pineapple_hater))
+	//list our ckey and assign it a favourite pizza
+	if(!pizza_preferences[nommer.ckey])
+		if(nommer.has_quirk(/datum/quirk/pineapple_liker))
+			pizza_preferences[nommer.ckey] = /obj/item/food/pizza/pineapple
+		else if(nommer.has_quirk(/datum/quirk/pineapple_hater))
 			var/list/pineapple_pizza_liker = pizza_types.Copy()
 			pineapple_pizza_liker -= /obj/item/food/pizza/pineapple
-			pizza_preferences[noms.ckey] = pickweight(pineapple_pizza_liker)
-		else if(noms.mind && noms.mind.assigned_role == "Botanist")
-			pizza_preferences[noms.ckey] = /obj/item/food/pizza/dank
-
-	var/obj/item/pizza_type = pizza_preferences[noms.ckey]
-	pizza = new pizza_type (src)
-	pizza.foodtypes = noms.dna.species.liked_food //it's our favorite!
+			pizza_preferences[nommer.ckey] = pickweight(pineapple_pizza_liker)
+		else if(nommer.mind && nommer.mind.assigned_role == "Botanist")
+			pizza_preferences[nommer.ckey] = /obj/item/food/pizza/dank
+		else
+			pizza_preferences[nommer.ckey] = pickweight(pizza_types)
+	if(pizza)
+		//if the pizza isn't our favourite, delete it
+		if(pizza.type != pizza_preferences[nommer.ckey])
+			QDEL_NULL(pizza)
+		else
+			pizza.foodtypes = nommer.dna.species.liked_food //make sure it's our favourite
+			return
+	if(pizza_preferences[nommer.ckey]) //make sure it hasn't been set to null, which happens sometimes
```

Comment:
This could result on a `null` `pizza` if `pickweight()` returns `null`, no?
The check should be put a bit before with an early return.

Reply:
It can result in a null pizza. I can early return but the result is an empty pizza box, which isn't optimal but isn't harmful.

Reply:
Correct me if I'm wrong, but doesn't it go this way?

- `if(!pizza_preferences[nommer.ckey])` is evaluated to false
- all the ifs inside of it evaluate to false, falling into the last one
- `pizza_preferences[nommer.ckey] = pickweight(pizza_types)` happens, but `pizza_type` has a length of zero by now
- `if(pizza)` is evaluated to true, and thus `if(pizza.type != pizza_preferences[nommer.ckey])` passes as the type is not -`null`
- `QDEL_NULL(pizza)` happens, and the box now has a `null` pizza
- `if(pizza_preferences[nommer.ckey])` evaluates to false, as this is `null
- now the box has no pizza

Reply:
Yeah, it can result in no pizza. So I added that check to prevent it trying to make a null pizza, which created a runtime.

Gameplay-wise, you open the box, it's empty and the incorrect boxtag can't be seen. Chances are extremely high closing and reopening it once will pick a pizza. But eventually closing and reopening it will fix it.

I can't think of a more optimal outcome (aside from fixing pick_weight() which I doubt I could figure out) and I don't understand why the early return would be better. Which is probably reflective of my coding knowledge.

The other thing I could do is loop up to three times if pick_weight() hasn't done its job to make it almost certain pick_weight() will work but I don't know if that would be acceptable in terms of performance and time taken to open it.

Reply:
First things first: why is `pizza_types` empty sometimes? Do you know that?
Or are you saying that `pickweight()` is returning `null` on a non-empty list?

Reply:
I don't know for certain but I'm pretty sure it's pickweight() causing the null. There's a comment on the second pickweight() proc (the one that allows null) mentioning that it happens sometimes.

Reply:
Add debugging lines when testing.
Things like
```dm
to_chat(world, "pizza_types length: [length(pizza_types)]")
```
Add `stack_traces()` if necessary.

Confirm `pickweight()` is outputting a `null` on a list with elements, and once you do, check what kind of contents the list has:
```dm
for(i in length(pizza_types))
	var/pizza_key = pizza_types[i]
	to_chat(world, "pizza_types [i]: [pizza_key] = [pizza_types[pizza_key]]")
```

Reply:
The code
```dm
for(var/i in 1 to 100)
	var/pizza_type_to_spawn = pickweight(pizza_types)
	if(pizza_type_to_spawn)
		to_chat(world, "pickweight() picked [pizza_type_to_spawn]!")
		//var/obj/item/food/pizza/spawned_pizza = new pizza_type_to_spawn(user.loc)
	else
		to_chat(world, "pickweight() returned null!")
		for(var/e in 1 to length(pizza_types))
			var/pizza_key = pizza_types[e]
			to_chat(world, "pizza_types [e]: [pizza_key] = [pizza_types[pizza_key]]")
```

The result

```dm
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/margherita!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/mushroom!
pickweight() picked /obj/item/food/pizza/meat!
pickweight() picked /obj/item/food/pizza/vegetable!
pickweight() picked /obj/item/food/pizza/pineapple!
pickweight() returned null!
pizza_types 1: /obj/item/food/pizza/meat = 1
pizza_types 2: /obj/item/food/pizza/mushroom = 1
pizza_types 3: /obj/item/food/pizza/margherita = 1
pizza_types 4: /obj/item/food/pizza/sassysage = 0.8
pizza_types 5: /obj/item/food/pizza/vegetable = 0.8
pizza_types 6: /obj/item/food/pizza/pineapple = 0.5
pizza_types 7: /obj/item/food/pizza/donkpocket = 0.3
pizza_types 8: /obj/item/food/pizza/dank = 0.1
```

Reply:
What I was suggesting to work around pick_weight() was something akin to the following:

```dm
/obj/item/pizzabox/floorpizza_spawner/attack_self(mob/living/user)
	for(var/i in 1 to 1000)
		var/pizza_type_to_spawn = pickweight(pizza_types)
		if(!pizza_type_to_spawn)
			for(var/e in 1 to 3)
				pizza_type_to_spawn = pickweight(pizza_types)
				if(pizza_type_to_spawn)
					break
		if(pizza_type_to_spawn)
			var/obj/item/food/pizza/spawned_pizza = new pizza_type_to_spawn(user.loc)
		else
			to_chat(world,"Loop [i] of 1000: No pizza type to spawn!")
```

I spawned 20,000 pizzas with this method and only two iterations ended with a null pizza type.

> Loop 852 of 1000: No pizza type to spawn!
> Loop 202 of 1000: No pizza type to spawn!

Reply:
I see the issue. `pickweight()` uses `rand()`, which uses integers, not floats.
Try multiplying each element of your list by 10 so you have no decimal values, and try again.
(ie: `/obj/item/food/pizza/pineapple = 5` instead of `= 0.5` )

Reply:
That seemed to fix it. Spawned another 20k pizzas without a single null.